From 990853eed4ec8fbad65d51531902ceef33b3e599 Mon Sep 17 00:00:00 2001
From: Suman Anna <s-anna@ti.com>
Date: Fri, 15 Feb 2019 12:28:35 -0600
Subject: [PATCH 40/51] soc: ti: pruss: Add support for PRU-ICSS subsystems on
 AM57xx SoCs

commit bc9dff3817b0a77df0993738ab1c2f2f1437e5f4 from
https://git.ti.com/cgit/processor-sdk/processor-sdk-linux/?h=processor-sdk-linux-01.00.00

The AM57xx family of SoCs supports two PRU-ICSS instances, each of
which has two PRU processor cores. The two PRU-ICSS instances are
identical to each other, and are very similar to the PRU-ICSS1 of
AM33xx/AM43xx except for a few minor differences like the RAM sizes
and the number of interrupts coming into the MPU INTC. They do
not have a programmable module reset line unlike those present on
AM33xx/AM43xx SoCs. The modules are reset just like any other IP
with the SoC's global cold/warm resets. Each PRU-ICSS's INTC is also
preceded by a Crossbar that enables multiple external events to be
routed to a specific number of input interrupt events. Any interrupt
event directed towards PRUSS needs this crossbar to be setup properly
on the firmware side.

The existing PRUSS platform drivers have been enhanced to support
these AM57xx PRU-ICSS instances through new AM57xx specific
compatibles for properly probing and booting all the different PRU
cores in each PRU-ICSS processor subsystem. The SoC-level integration
differences are dealt with using match data within the PRUSS SoC bus
driver. A build dependency with SOC_DRA7XX is also added to enable
the driver to be built in AM57xx-only configuration (there is no
separate Kconfig option for AM57xx vs DRA7xx). The initial names
for the firmware images for each PRU core are retrieved from DT
nodes, and can be adjusted through sysfs if required.

Signed-off-by: Suman Anna <s-anna@ti.com>
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 drivers/irqchip/irq-pruss-intc.c | 4 ++++
 drivers/remoteproc/pru_rproc.c   | 1 +
 drivers/soc/ti/pruss.c           | 1 +
 drivers/soc/ti/pruss_soc_bus.c   | 5 +++++
 4 files changed, 11 insertions(+)

diff --git a/drivers/irqchip/irq-pruss-intc.c b/drivers/irqchip/irq-pruss-intc.c
index 5885ccb44423..022b4cfa811a 100644
--- a/drivers/irqchip/irq-pruss-intc.c
+++ b/drivers/irqchip/irq-pruss-intc.c
@@ -613,6 +613,10 @@ static const struct of_device_id pruss_intc_of_match[] = {
 		.compatible = "ti,am4376-pruss-intc",
 		.data = &am437x_pruss_intc_data,
 	},
+	{
+		.compatible = "ti,am5728-pruss-intc",
+		.data = NULL,
+	},
 	{
 		.compatible = "ti,k2g-pruss-intc",
 		.data = &k2g_pruss_intc_data,
diff --git a/drivers/remoteproc/pru_rproc.c b/drivers/remoteproc/pru_rproc.c
index 8339ca960e13..af53b975f7aa 100644
--- a/drivers/remoteproc/pru_rproc.c
+++ b/drivers/remoteproc/pru_rproc.c
@@ -594,6 +594,7 @@ static int pru_rproc_remove(struct platform_device *pdev)
 static const struct of_device_id pru_rproc_match[] = {
 	{ .compatible = "ti,am3356-pru", },
 	{ .compatible = "ti,am4376-pru", },
+	{ .compatible = "ti,am5728-pru", },
 	{ .compatible = "ti,k2g-pru",    },
 	{},
 };
diff --git a/drivers/soc/ti/pruss.c b/drivers/soc/ti/pruss.c
index 51ef9c4746d0..656739b4a204 100644
--- a/drivers/soc/ti/pruss.c
+++ b/drivers/soc/ti/pruss.c
@@ -202,6 +202,7 @@ static const struct pruss_match_private_data am437x_match_data[] = {
 static const struct of_device_id pruss_of_match[] = {
 	{ .compatible = "ti,am3356-pruss", .data = NULL, },
 	{ .compatible = "ti,am4376-pruss", .data = &am437x_match_data, },
+	{ .compatible = "ti,am5728-pruss", .data = NULL, },
 	{ .compatible = "ti,k2g-pruss", .data = NULL, },
 	{ /* sentinel */ },
 };
diff --git a/drivers/soc/ti/pruss_soc_bus.c b/drivers/soc/ti/pruss_soc_bus.c
index 46f72c4fa61b..f7bb75753c77 100644
--- a/drivers/soc/ti/pruss_soc_bus.c
+++ b/drivers/soc/ti/pruss_soc_bus.c
@@ -257,9 +257,14 @@ static const struct pruss_soc_bus_match_data am437x_data = {
 	.has_reset = true,
 };
 
+static const struct pruss_soc_bus_match_data am57xx_data = {
+	.has_reset = false,
+};
+
 static const struct of_device_id pruss_soc_bus_of_match[] = {
 	{ .compatible = "ti,am3356-pruss-soc-bus", .data = &am335x_data, },
 	{ .compatible = "ti,am4376-pruss-soc-bus", .data = &am437x_data, },
+	{ .compatible = "ti,am5728-pruss-soc-bus", .data = &am57xx_data, },
 	{ /* sentinel */ },
 };
 MODULE_DEVICE_TABLE(of, pruss_soc_bus_of_match);
-- 
2.17.1

